src/webgpu/api/validation/resource_usages/textureUsageInRender.spec.ts

```diff
@@ -173,3 +179,59 @@ g.test('readwrite_upon_aspects')
       encoder.finish();
     }, !success);
   });
+
+g.test('shader_stages_and_visibility')
+  .params(
+    params()
+      .combine(poptions('readVisibility', kShaderStageCombinations))
+      .combine(poptions('writeVisibility', [0, ...kShaderStages]))
+  )
+  .fn(async t => {
+    const { readVisibility, writeVisibility } = t.params;
+
+    // writeonly-storage-texture binding type is not supported in vertex stage. So, this test
+    // uses writeonly-storage-texture binding as writable binding upon the same subresource if
+    // vertex stage is not included. Otherwise, it uses output attachment instead.
+    const writeHasVertexStage = writeVisibility & GPUShaderStage.VERTEX;
+    const texUsage = writeHasVertexStage
+      ? GPUTextureUsage.SAMPLED | GPUTextureUsage.OUTPUT_ATTACHMENT
+      : GPUTextureUsage.SAMPLED | GPUTextureUsage.STORAGE;
+
+    const texture = t.createTexture({ usage: texUsage });
+    const view = texture.createView();
+    const bglEntries: GPUBindGroupLayoutEntry[] = [
+      { binding: 0, visibility: readVisibility, type: 'sampled-texture' },
+    ];
+    const bgEntries: GPUBindGroupEntry[] = [{ binding: 0, resource: view }];
+    if (!writeHasVertexStage) {
+      bglEntries.push({
+        binding: 1,
+        visibility: writeVisibility,
+        type: 'writeonly-storage-texture',
+        storageTextureFormat: 'rgba8unorm',
+      });
+      bgEntries.push({ binding: 1, resource: view });
+    }
+    const bindGroup = t.device.createBindGroup({
+      entries: bgEntries,
+      layout: t.device.createBindGroupLayout({ entries: bglEntries }),
+    });
+
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: writeHasVertexStage ? view : t.createTexture().createView(),
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+    pass.setBindGroup(0, bindGroup);
+    pass.endPass();
+
+    // Texture usages in bindings with invisible shader stages should be tracked.
```

Comment:
nit: this test isn't super clear about what is meant be "invisible shader stages". Maybe we should say "with visibility none" ?

Reply:
Yeah. I explained a little bit in the comment about invisible shader stages.